File: rust/src/endpoints/bankid_no.rs
use crate::client::{IdkollenError, PollOptions, WaitError};
use crate::models::{AgeVerificationRequest, AgeVerificationStatus};
use crate::models::{
    BankIdNoAuthRequest, BankIdNoBackchannelAuthRequest, BankIdNoSignRequest, BankIdNoStatus,
};
use std::time::Instant;

#[cfg(feature = "async")]
use crate::client::IdkollenClient;

#[cfg(feature = "async")]
pub struct BankIdNoEndpoint<'a>(pub(crate) &'a IdkollenClient);

#[cfg(feature = "async")]
impl BankIdNoEndpoint<'_> {
    /// Start a BankID NO authentication session.
    pub async fn auth(&self, req: BankIdNoAuthRequest) -> Result<BankIdNoStatus, IdkollenError> {
        self.0.post("/v3/bankid-no/auth", &req).await
    }

    /// Start a BankID NO backchannel authentication session.
    pub async fn backchannel_auth(
        &self,
        req: BankIdNoBackchannelAuthRequest,
    ) -> Result<BankIdNoStatus, IdkollenError> {
        self.0.post("/v3/bankid-no/backchannel/auth", &req).await
    }

    /// Start a BankID NO signing session.
    pub async fn sign(&self, req: BankIdNoSignRequest) -> Result<BankIdNoStatus, IdkollenError> {
        self.0.post("/v3/bankid-no/sign", &req).await
    }

    /// Start a BankID NO age verification session.
    pub async fn age_verification(
        &self,
        req: AgeVerificationRequest,
    ) -> Result<AgeVerificationStatus, IdkollenError> {
        self.0.post("/v3/bankid-no/age-verification", &req).await
    }

    /// Poll the current status of a BankID NO authentication session.
    pub async fn auth_status(&self, id: &str) -> Result<BankIdNoStatus, IdkollenError> {
        self.0.get(&format!("/v3/bankid-no/auth/{}", id)).await
    }

    /// Poll the current status of a BankID NO signing session.
    pub async fn sign_status(&self, id: &str) -> Result<BankIdNoStatus, IdkollenError> {
        self.0.get(&format!("/v3/bankid-no/sign/{}", id)).await
    }

    /// Poll the current status of a BankID NO age verification session.
    pub async fn age_verification_status(
        &self,
        id: &str,
    ) -> Result<AgeVerificationStatus, IdkollenError> {
        self.0
            .get(&format!("/v3/bankid-no/age-verification/{}", id))
            .await
    }

    /// Cancel a BankID NO authentication session.
    pub async fn cancel_auth(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/bankid-no/auth/{}", id)).await
    }

    /// Cancel a BankID NO signing session.
    pub async fn cancel_sign(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/bankid-no/sign/{}", id)).await
    }

    /// Cancel a BankID NO age verification session.
    pub async fn cancel_age_verification(&self, id: &str) -> Result<(), IdkollenError> {
        self.0
            .delete(&format!("/v3/bankid-no/age-verification/{}", id))
            .await
    }

    /// Poll until the authentication session reaches a terminal state or the timeout elapses.
    pub async fn wait_for_auth(
        &self,
        id: &str,
        opts: PollOptions,
    ) -> Result<BankIdNoStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.auth_status(id).await?;

            match status {
                BankIdNoStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    tokio::time::sleep(opts.interval).await;
                },
                terminal => return Ok(terminal),
            }
        }
    }

    /// Poll until the signing session reaches a terminal state or the timeout elapses.
    pub async fn wait_for_sign(
        &self,
        id: &str,
        opts: PollOptions,
    ) -> Result<BankIdNoStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.sign_status(id).await?;

            match status {
                BankIdNoStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    tokio::time::sleep(opts.interval).await;
                },
                terminal => return Ok(terminal),
            }
        }
    }

    /// Poll until the age verification session reaches a terminal state or the timeout elapses.
    pub async fn wait_for_age_verification(
        &self,
        id: &str,
        opts: PollOptions,
    ) -> Result<AgeVerificationStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.age_verification_status(id).await?;

            match status {
                AgeVerificationStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    tokio::time::sleep(opts.interval).await;
                },
                terminal => return Ok(terminal),
            }
        }
    }
}

#[cfg(feature = "blocking")]
use crate::client::IdkollenBlockingClient;

#[cfg(feature = "blocking")]
pub struct BankIdNoBlockingEndpoint<'a>(pub(crate) &'a IdkollenBlockingClient);

#[cfg(feature = "blocking")]
impl BankIdNoBlockingEndpoint<'_> {
    /// Start a BankID NO authentication session.
    pub fn auth(&self, req: BankIdNoAuthRequest) -> Result<BankIdNoStatus, IdkollenError> {
        self.0.post("/v3/bankid-no/auth", &req)
    }

    /// Start a BankID NO backchannel authentication session.
    pub fn backchannel_auth(
        &self,
        req: BankIdNoBackchannelAuthRequest,
    ) -> Result<BankIdNoStatus, IdkollenError> {
        self.0.post("/v3/bankid-no/backchannel/auth", &req)
    }

    /// Start a BankID NO signing session.
    pub fn sign(&self, req: BankIdNoSignRequest) -> Result<BankIdNoStatus, IdkollenError> {
        self.0.post("/v3/bankid-no/sign", &req)
    }

    /// Start a BankID NO age verification session.
    pub fn age_verification(
        &self,
        req: AgeVerificationRequest,
    ) -> Result<AgeVerificationStatus, IdkollenError> {
        self.0.post("/v3/bankid-no/age-verification", &req)
    }

    /// Poll the current status of a BankID NO authentication session.
    pub fn auth_status(&self, id: &str) -> Result<BankIdNoStatus, IdkollenError> {
        self.0.get(&format!("/v3/bankid-no/auth/{}", id))
    }

    /// Poll the current status of a BankID NO signing session.
    pub fn sign_status(&self, id: &str) -> Result<BankIdNoStatus, IdkollenError> {
        self.0.get(&format!("/v3/bankid-no/sign/{}", id))
    }

    /// Poll the current status of a BankID NO age verification session.
    pub fn age_verification_status(
        &self,
        id: &str,
    ) -> Result<AgeVerificationStatus, IdkollenError> {
        self.0
            .get(&format!("/v3/bankid-no/age-verification/{}", id))
    }

    /// Cancel a BankID NO authentication session.
    pub fn cancel_auth(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/bankid-no/auth/{}", id))
    }

    /// Cancel a BankID NO signing session.
    pub fn cancel_sign(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/bankid-no/sign/{}", id))
    }

    /// Cancel a BankID NO age verification session.
    pub fn cancel_age_verification(&self, id: &str) -> Result<(), IdkollenError> {
        self.0
            .delete(&format!("/v3/bankid-no/age-verification/{}", id))
    }

    /// Poll until the authentication session reaches a terminal state or the timeout elapses.
    pub fn wait_for_auth(&self, id: &str, opts: PollOptions) -> Result<BankIdNoStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.auth_status(id)?;

            match status {
                BankIdNoStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    std::thread::sleep(opts.interval);
                },
                terminal => return Ok(terminal),
            }
        }
    }

    /// Poll until the signing session reaches a terminal state or the timeout elapses.
    pub fn wait_for_sign(&self, id: &str, opts: PollOptions) -> Result<BankIdNoStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.sign_status(id)?;

            match status {
                BankIdNoStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    std::thread::sleep(opts.interval);
                },
                terminal => return Ok(terminal),
            }
        }
    }

    /// Poll until the age verification session reaches a terminal state or the timeout elapses.
    pub fn wait_for_age_verification(
        &self,
        id: &str,
        opts: PollOptions,
    ) -> Result<AgeVerificationStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.age_verification_status(id)?;

            match status {
                AgeVerificationStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    std::thread::sleep(opts.interval);
                },
                terminal => return Ok(terminal),
            }
        }
    }
}

File: js/src/endpoints/bankid_no.ts
import { IdkollenClient, IdkollenError, PollOptions } from "@/client";
import type {
  BankIdNoAuthRequest,
  BankIdNoBackchannelAuthRequest,
  BankIdNoSignRequest,
  BankIdNoStatus,
} from "@/models/bankid_no";
import type { AgeVerificationRequest, AgeVerificationStatus } from "@/models/age_verification";

const sleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

/** BankID NO operations. */
export class BankIdNoEndpoint {
  constructor(private readonly client: IdkollenClient) {}

  /** Start a BankID NO authentication session. */
  public async auth(req: BankIdNoAuthRequest): Promise<BankIdNoStatus> {
    return this.client._post("/v3/bankid-no/auth", req);
  }

  /** Start a BankID NO backchannel authentication session. */
  public async backchannelAuth(req: BankIdNoBackchannelAuthRequest): Promise<BankIdNoStatus> {
    return this.client._post("/v3/bankid-no/backchannel/auth", req);
  }

  /** Start a BankID NO signing session. */
  public async sign(req: BankIdNoSignRequest): Promise<BankIdNoStatus> {
    return this.client._post("/v3/bankid-no/sign", req);
  }

  /** Poll the current status of a BankID NO authentication session. */
  public async authStatus(id: string): Promise<BankIdNoStatus> {
    return this.client._get(`/v3/bankid-no/auth/${id}`);
  }

  /** Poll the current status of a BankID NO signing session. */
  public async signStatus(id: string): Promise<BankIdNoStatus> {
    return this.client._get(`/v3/bankid-no/sign/${id}`);
  }

  /** Cancel a BankID NO authentication session. */
  public async cancelAuth(id: string): Promise<void> {
    return this.client._delete(`/v3/bankid-no/auth/${id}`);
  }

  /** Cancel a BankID NO signing session. */
  public async cancelSign(id: string): Promise<void> {
    return this.client._delete(`/v3/bankid-no/sign/${id}`);
  }

  /** Start a BankID NO age verification session. */
  public async ageVerification(req: AgeVerificationRequest): Promise<AgeVerificationStatus> {
    return this.client._post("/v3/bankid-no/age-verification", req);
  }

  /** Poll the current status of a BankID NO age verification session. */
  public async ageVerificationStatus(id: string): Promise<AgeVerificationStatus> {
    return this.client._get(`/v3/bankid-no/age-verification/${id}`);
  }

  /** Cancel a BankID NO age verification session. */
  public async cancelAgeVerification(id: string): Promise<void> {
    return this.client._delete(`/v3/bankid-no/age-verification/${id}`);
  }

  /**
   * Poll until the authentication session reaches a terminal state or the timeout elapses.
   * @throws {IdkollenError} with code `"poll_timeout"` if the timeout is exceeded.
   */
  public async waitForAuth(
    id: string,
    opts: PollOptions = new PollOptions(),
  ): Promise<BankIdNoStatus> {
    const deadline = Date.now() + opts.timeoutMs;

    while (true) {
      const status = await this.authStatus(id);

      if (status.status !== "PENDING") {
        return status;
      }

      if (Date.now() >= deadline) {
        throw new IdkollenError("poll_timeout", 0, "Poll timed out");
      }

      await sleep(opts.intervalMs);
    }
  }

  /**
   * Poll until the signing session reaches a terminal state or the timeout elapses.
   * @throws {IdkollenError} with code `"poll_timeout"` if the timeout is exceeded.
   */
  public async waitForSign(
    id: string,
    opts: PollOptions = new PollOptions(),
  ): Promise<BankIdNoStatus> {
    const deadline = Date.now() + opts.timeoutMs;

    while (true) {
      const status = await this.signStatus(id);

      if (status.status !== "PENDING") {
        return status;
      }

      if (Date.now() >= deadline) {
        throw new IdkollenError("poll_timeout", 0, "Poll timed out");
      }

      await sleep(opts.intervalMs);
    }
  }

  /**
   * Poll until the age verification session reaches a terminal state or the timeout elapses.
   * @throws {IdkollenError} with code `"poll_timeout"` if the timeout is exceeded.
   */
  public async waitForAgeVerification(
    id: string,
    opts: PollOptions = new PollOptions(),
  ): Promise<AgeVerificationStatus> {
    const deadline = Date.now() + opts.timeoutMs;

    while (true) {
      const status = await this.ageVerificationStatus(id);

      if (status.status !== "PENDING") {
        return status;
      }

      if (Date.now() >= deadline) {
        throw new IdkollenError("poll_timeout", 0, "Poll timed out");
      }

      await sleep(opts.intervalMs);
    }
  }
}
